src/index.ts

```diff
@@ -1,43 +1,145 @@
-  const srcOptions = scopedOptions(scope, from.registry, from.token)
-  // fullMetadata may needed to obtain the repository property in manifest
-  srcOptions.fullMetadata = true
-  const srcPackument = await getPackument(name, srcOptions)
-  const srcVersions = srcPackument ? Object.keys(srcPackument.versions) : []
-  console.debug('Source versions', srcVersions)
-
-  // get available target versions
-  const dstOptions = scopedOptions(scope, to.registry, to.token)
-  const dstPackument = await getPackument(name, dstOptions)
-  const dstVersions = dstPackument ? Object.keys(dstPackument.versions) : []
-  console.debug('Target versions', dstVersions)
+import {ensureRepositoryAccess, fetchTarball, getScopedOptions, prepareManifest, publish} from './utils'
+import * as pacote from "pacote"
+import {filterSourceVersions} from "./filters"
+import {MigrationSettings} from "./settings"
+
+const logger = require('pino')()
+
+export async function syncPackages(settings: MigrationSettings): Promise<number> {
+  let result = 0
 
-  const missing = srcVersions.filter(x => !dstVersions.includes(x))
-  console.log('Missing versions', missing)
-  if (!missing.length) {
+  if (!await ensureRepositoryAccess(settings)) {
     return 0
   }
 
-  for (const version of missing) {
-    const spec = name + '@' + version
-    console.log('Reading %s from %s', spec, from.registry)
+  for (let i = 0; i < settings.packages.length; i++) {
+    const npmPackage = settings.packages[i]
+    logger.info(`[${npmPackage}] synchronisation started`)
+    let versionsImported = await syncPackage(settings, npmPackage)
+    logger.info(`[${npmPackage}] imported ${versionsImported} versions`)
+    logger.info(`--------------------------`)
 
-    const manifest = prepareManifest(srcPackument, version)
-    console.debug('Dist', manifest.dist)
+    result += versionsImported
+  }
+  return result
+}
 
-    //const tarball = await getTarball(spec, srcOptions)
-    const tarball = await fetchTarball(manifest.dist, from.token)
-    console.debug('Tarball length', tarball.length)
+async function syncPackage(
+  settings: MigrationSettings,
+  packageName: string
+) {
+  // fullMetadata may be needed to obtain the repository property in manifest
+  const srcPackument = await getPackageMetadata(settings.source, packageName, true)
+  const srcVersions = filterSourceVersions(packageName, srcPackument, settings.migrationMode);
+  logger.debug(`[${packageName}] pre-selected source version: ${srcVersions}`)
 
-    console.log('Publishing %s to %s', spec, to.registry)
-    await publish(manifest, tarball, dstOptions, dryRun)
+  if (srcVersions.length === 0) {
+    logger.info(`[${packageName}] nothing to sync`)
+    return 0
   }
-  return missing.length
+
+  const dstPackument = await getPackageMetadata(settings.target, packageName, false)
+  const dstVersions = dstPackument ? Object.keys(dstPackument.versions) : []
+  logger.debug(`[${packageName}] found target versions: ${dstVersions}`)
```

Comment:
Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/joebowbeer/regsync/pullRequest?prid=9263162)
```suggestion
  logger.debug(`[${packageName}] found target versions: ${dstVersions}`);
```